### DISCUSSION TITLE
补充`laravel`在`s6`环境下调优

### DISCUSSION THREAD

**Discussion by cgfeel** [khs1994-docker/lnmp#1234] [General]
分别在不同环境的配置文件下指向不同目录位置，以下`s6`配置文件都是基于这个目录之下
- 开发环境数据卷：`./s6-overlay-develop`
- 线上环境数据卷：`./s6-overlay`

关于`laravel`缓存，线上和线下要反着来（在`./scripts/laravel-prepare`）：
线上创建缓存
```
php /app/api_service/artisan config:cache
php /app/api_service/artisan route:cache
php /app/api_service/artisan view:cache
php /app/api_service/artisan queue:restart
```
线下清除缓存
```
php /app/api_service/artisan config:clear
php /app/api_service/artisan route:clear
php /app/api_service/artisan view:clear
php /app/api_service/artisan queue:restart
```

---

如果线下开发有自建证书：那么还要将开发环境中安装CA的支持，并将hosts添加到容器（在`./scripts/laravel-prepare`）：
```
apk --no-cache add ca-certificates && update-ca-certificates
echo '192.168.31.198    api.local.domian.com' >> /etc/hosts
```

在配置文件`docker-lnmp.override.yml`中，将数据卷挂上
```
      - /home/levi/app/config/nginx/ssl-self/root-ca.crt:/usr/local/share/ca-certificates/root-ca.crt:ro
```

然后将`php.ini`中ca的目录位置分别为`curl`和`openssl`加上：
```
curl.capath=/usr/local/share/ca-certificates
openssl.capath=/usr/local/share/ca-certificates
```

---

在线上开发环境默认是没有打开`opcache`的
 - 既不是配置`php.ini`，也不是配置`php-ext-opcache.ini`
 - 而是打开环境打开容器环境变量`LNMP_OPCACHE_ENABLE=1`，如果要调整额外配置，可以修改`docker-php.example.ini`

详细见：https://github.com/khs1994-docker/lnmp/issues/1233

**Comment by cgfeel** [comment]
再补充一点，laravel中tinker在容器中无法正常使用，会报错：
```
/usr/bin/less: unrecognized option: X
```

由于镜像没有提供，所以还是在`s6`下进行解决（在`./scripts/laravel-prepare`）：
```
apk add --update --no-cache less  
```

参考：
 - https://qiita.com/www_y118/items/c473907836db13bd42ca
 - https://github.com/deivid-rodriguez/pry-byebug/issues/368
 - https://github.com/laravel/tinker/issues/52

**Comment by cgfeel** [comment]
在补充一个开发环境优化，在`vscode`中开发`php`
 - 假定宿主机是干净的，没有安装过`php`
 - 那么要针对容器`php8`进行开发，会提示找不到可运行的路径

解决办法，这个时候在`/usr/local/bin/php`下添加一个脚本
```
result=$(/usr/bin/docker container ls --format '{{.ID}}' -f label=com.khs1994.lnmp -f label=com.docker.compose.service=php8)
if [ ! $result ]; then
  echo "container is empty"; exit 1;
fi

command="docker exec ${result} php "$@""
$command
```

然后打开`vscode`，在设置中，如果是远程开发就在`remote`设置中，搜索php，然后选择`Edit in setting.json`，把这段加进去：
```
"php.validate.executablePath": "/usr/local/bin/php"
```

**Comment by cgfeel** [comment]
补充一个快速查找`artisan`的`shell`：
```
#!/bin/bash

if [[ 'local' == $1 ]]
then
  result=$(/usr/bin/docker container ls --format '{{.ID}}' -f label=com.khs1994.lnmp -f label=com.docker.compose.service=php8)
  EXEC_PARAMS=${@:2:$#}
else
  result=$(/usr/bin/docker container ls --format '{{.ID}}' -f label=com.khs1994.lnmp -f label=com.docker.swarm.service.name=lnmp_php8)
  EXEC_PARAMS=$@
fi

if [ ! $result ]; then
  echo "container is empty"; exit 1;
fi

/usr/bin/docker exec -it $result sh -c "php /app/api_service/artisan $EXEC_PARAMS"
```
这样线上只要执行`artisan.run`就可以调用容器的`artisan`，开发环境通过`artisan.run local`

---

这样还是存在一个问题，通过`shell`脚本执行`artisan`后创建的相关文件所属用户是`root:root`，比如：
```
./cli/artisan.run local make:listener Gold/TaskGoldListener
```
如果本地远程开发并非`root`，那么就存在不能编辑的问题，这样就需要手动再次：`sudo chown {user}:{group} {path}`
对于这个问题我暂时没有好的解决办法。。。。

**Comment by cgfeel** [comment]
补充一个本地开发相关，和`laravel`关系不大，`webpack`在容器下报错
```
Watchpack Error (watcher): Error: ENOSPC: System limit for number of file watchers reached, watch
```
解决办法参考：https://github.com/gatsbyjs/gatsby/issues/11406#issuecomment-458769756

**Comment by cgfeel** [comment]
再补充一个移动端本地开发技巧，将自签名ca证书导入到手机，Android见：https://blog.csdn.net/weixin_43856764/article/details/120442011
然后修改本地路由的hosts，我用的是小米路由，可以自定义路由的hosts，将域名的host指向容器宿主机的内网IP，这样就可以顺利请求本地开发环境了。

- 将线上的域名加个`local`作为本地开发环境域名，通过自签名进行访问，能够实现和线上一样的效果
- 服务端：不同的环境下通过项目的`.env`来确定请求指向的URL
- 客户端：修改hosts指向开发环境专属的URL

例如像这样：
```
192.168.31.198   local.migdat.com
192.168.31.198    admin.local.migdat.com
192.168.31.198    api.local.migdat.com
```

**Comment by cgfeel** [comment]
修正上面几处问题
```
apk --no-cache add ca-certificates && update-ca-certificates
echo '192.168.31.198    api.local.domian.com' >> /etc/hosts
```
host可以在compose里加上
```
    extra_hosts:
      - "api.local.migdat.com:192.168.31.198"
```
apt不能加到`./scripts/laravel-prepare`中，否则会应为启动时间过长导致失败
```
fetch https://dl-cdn.alpinelinux.org/alpine/v3.15/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.15/community/x86_64/APKINDEX.tar.gz
(1/1) Installing less (590-r0)
Executing busybox-1.34.1-r3.trigger
OK: 78 MiB in 78 packages
fetch https://dl-cdn.alpinelinux.org/alpine/v3.15/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.15/community/x86_64/APKINDEX.tar.gz
s6-rc: fatal: timed out
s6-sudoc: fatal: unable to get exit status from server: Operation timed out
OK: 78 MiB in 78 packages
```

目前我是开发环境启动后手动执行，如果要策底解决，可能需要重做镜像。。。。
```
docker exec {id} apk add --update --no-cache less
docker exec {id} apk apk --no-cache add ca-certificates
docker exec {id} update-ca-certificates
```

**Reply by khs1994** [reply]
需要增加包最好基于镜像自己构建

**Reply by khs1994** [reply]
>否则会应为启动时间过长导致失败

因为 alpine apk 源 dl-cdn.alpinelinux.org 在国内访问特别慢，换国内源会很快

**Reply by cgfeel** [reply]
> > 否则会应为启动时间过长导致失败
> 
> 因为 alpine apk 源 dl-cdn.alpinelinux.org 在国内访问特别慢，换国内源会很快

好的，我试试，因为只是本地开发环境要用到，所以也没有做镜像
